pkg/scheduler/framework/plugins/interpodaffinity/scoring.go

```diff
@@ -144,6 +168,10 @@ func (pl *InterPodAffinity) PreScore(
 	hasAffinityConstraints := affinity != nil && affinity.PodAffinity != nil
 	hasAntiAffinityConstraints := affinity != nil && affinity.PodAntiAffinity != nil
 
+	if err := validatePreferredAffinityTerms(pod); err != nil {
```

Comment:
Is it ok to check the podInfo.PreferredAffinityTerms and podInfo.PreferredAntiAffinityTerms is nil or not in line 195? Since the NewPodInfo already did the same check, then we can reuse that result.

Reply:
@soulxu using the check that's done in `NewPodInfo` was actually the other option I considered in https://github.com/kubernetes/kubernetes/pull/92714#issuecomment-667227790 for both PreScore and PreFilter. The tradeoff with checking it there is that `NewPodInfo` would then need to add error handling everywhere it is called (which is quite a few places).

So while this approach is redundant, I think it is clearer what is being checked and does it as early as possible. By making the check distinct we can also ensure that it stays consistent between versions (as various refactors have already changed how this error is handled, see my main comment in #92714)

Of course if people would prefer that we use the check in `NewPodInfo` I am open to doing it that way as well and can make a PR with that too

Reply:
Oh I see, or we can just get the AffinityTerms lists *from* the PodInfo returned by NewPodInfo. Yeah that could save us having to do the check again I agree

Reply:
> Oh I see, or we can just get the AffinityTerms lists _from_ the PodInfo returned by NewPodInfo. Yeah that could save us having to do the check again I agree

Yea

Reply:
Actually, that won't work because if NewPodInfo hits an error parsing the label selector, it just throws out the error and returns nil so we wouldn't be able to actually check that selector. Meaning we would still have to bubble the error up like in https://github.com/kubernetes/kubernetes/pull/92714#issuecomment-667227790

Reply:
how about check like this: "pod.Spec.Affinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution != nil and podInfo.PreferredAffinityTerms == nil", then we know the pod has preferred affinity but newPodInfo return nil, it means there is invalid affinity.

Reply:
fine with me. The point is avoiding parsing, which has proven to be costly. Although it's acceptable one per scheduling cycle.

Reply:
> The point is avoiding parsing, which has proven to be costly

Are you referring to the `metav1.LabelSelectorAsSelector` call? It would still be good to call that in order to get the error, or in that case lean toward storing it in PodInfo like you said in https://github.com/kubernetes/kubernetes/pull/93660#discussion_r465066386

Reply:
your call